### ISSUE TITLE
testlimit.lua fails with "yield across metamethod/C-call boundary"

### ISSUE THREAD

**Issue by hishamhm** [keplerproject/copas#35]
I tried running the `testlimit.lua` test with both Copas 2.0.0 and the current Git HEAD (bd3de0b), using both Lua 5.1 and 5.3, and got the same output:

```
lua-5.1 ./testlimit.lua                         
Starting: http://www.google.com...
...Users/hisham/.luarocks/share/lua/5.1/copas/limit.lua:21: attempt to yield across metamethod/C-call boundary	thread: 0x8148dc0	nil
Starting: http://www.microsoft.com...
...Users/hisham/.luarocks/share/lua/5.1/copas/limit.lua:21: attempt to yield across metamethod/C-call boundary	thread: 0x8149258	nil
Starting: http://www.apple.com...
...Users/hisham/.luarocks/share/lua/5.1/copas/limit.lua:21: attempt to yield across metamethod/C-call boundary	thread: 0x8149730	nil
Starting: http://www.facebook.com...
...Users/hisham/.luarocks/share/lua/5.1/copas/limit.lua:21: attempt to yield across metamethod/C-call boundary	thread: 0x8149c50	nil
Starting: http://www.yahoo.com...
...Users/hisham/.luarocks/share/lua/5.1/copas/limit.lua:21: attempt to yield across metamethod/C-call boundary	thread: 0x814a0a0	nil
All tasks finished
I was also waiting...
```

These are the rocks I have installed:

```
Installed rocks:
----------------

alt-getopt
   0.7.0-1 (installed) - /Users/hisham/.luarocks/lib/luarocks/rocks-5.1

copas
   2.0.0-1 (installed) - /Users/hisham/.luarocks/lib/luarocks/rocks-5.1

coxpcall
   1.15.0-1 (installed) - /Users/hisham/.luarocks/lib/luarocks/rocks-5.1

dkjson
   2.5-2 (installed) - /Users/hisham/.luarocks/lib/luarocks/rocks-5.1

lpeg
   1.0.0-1 (installed) - /Users/hisham/.luarocks/lib/luarocks/rocks-5.1

luafilesystem
   1.6.3-1 (installed) - /Users/hisham/.luarocks/lib/luarocks/rocks-5.1

luarocks
   2.2.3rc1-1 (installed) - /Users/hisham/.luarocks/lib/luarocks/rocks-5.1

luasec
   scm-1 (installed) - /Users/hisham/.luarocks/lib/luarocks/rocks-5.1

luasocket
   3.0rc1-2 (installed) - /Users/hisham/.luarocks/lib/luarocks/rocks-5.1

moonscript
   0.3.2-1 (installed) - /Users/hisham/.luarocks/lib/luarocks/rocks-5.1
```

Any idea what's going on?

**Comment by mpeterv** [comment]
I think I've fixed this by installing luasocket from its git HEAD as well

**Comment by Tieske** [comment]
I had the same failure with Lua 5.1, but it works with LuaJIT.

**Comment by Tieske** [comment]
with the latest luasocket (scm rockspec from the github repo) I still get the error (on Lua 5.1)

**Comment by Tieske** [comment]
`limit.lua` didn't use coxpcall on Lua 5.1, which was the source of the error. Now fixed in master.

**Comment by Tieske** [comment]
@hishamhm  rereading the OP, are you sure you got the same error with 5.3?

**Comment by mpeterv** [comment]
Copas before fix (#38), Luasocket 3.0rc1-2, Lua 5.1: error
Copas before fix, Luasocket 3.0rc1-2, Lua 5.3: error
Copas before fix, Luasocket master, Lua 5.1: error
Copas before fix, Luasocket master, Lua 5.3: OK
Copas after fix, Luasocket 3.0rc1-2, Lua 5.1: OK
Copas after fix, Luasocket 3.0rc1-2, Lua 5.3: error
Copas after fix, Luasocket master, Lua 5.1: OK
Copas after fix , Luasocket master, Lua 5.3: OK

**Comment by mpeterv** [comment]
Is it possible to fix error on Lua 5.3 + Luasocket 3.0 or is there a related problem in Luasocket?

**Comment by mpeterv** [comment]
Right, there was a problem in Luasocket that breaks socket.protect for Lua 5.2 and 5.3, fixed [here](https://github.com/diegonehab/luasocket/commit/0b03eec). More than a year passed and there is no patch release.
